analytic_engine/src/sst/parquet/encoding.rs

```diff
@@ -150,3 +189,269 @@ pub fn decode_sst_meta_data(kv: &KeyValue) -> Result<SstMetaData> {
+    fn encode(&mut self, arrow_record_batch_vec: Vec<ArrowRecordBatch>) -> Result<usize> {
+        assert!(self.arrow_writer.is_some());
+
+        let record_batch = ArrowRecordBatch::concat(&self.arrow_schema, &arrow_record_batch_vec)
+            .map_err(|e| Box::new(e) as _)
+            .context(EncodeRecordBatch)?;
+
+        self.arrow_writer
+            .as_mut()
+            .unwrap()
+            .write(&record_batch)
+            .map_err(|e| Box::new(e) as _)
+            .context(EncodeRecordBatch)?;
+
+        Ok(record_batch.num_rows())
+    }
+
+    fn close(&mut self) -> Result<Vec<u8>> {
+        assert!(self.arrow_writer.is_some());
+
+        let arrow_writer = self.arrow_writer.take().unwrap();
+        arrow_writer
+            .close()
+            .map_err(|e| Box::new(e) as _)
+            .context(EncodeRecordBatch)?;
+
+        Ok(self.buf.clone().into_bytes())
+    }
+}
+
+struct HybridRecordEncoder {
+    buf: EncodingWriter,
+    // wrap in Option so ownership can be take out behind `&mut self`
+    arrow_writer: Option<ArrowWriter<EncodingWriter>>,
+    arrow_schema: ArrowSchemaRef,
+    tsid_type: IndexedType,
+    timestamp_type: IndexedType,
+    key_types: Vec<IndexedType>,
+    // columns that can be collpased into list
+    non_key_types: Vec<IndexedType>,
+}
+
+impl HybridRecordEncoder {
+    fn try_new(write_props: WriterProperties, schema: &Schema) -> Result<Self> {
+        let tsid_idx = schema.index_of_tsid().context(TsidRequired)?;
+        let timestamp_type = IndexedType {
+            idx: schema.timestamp_index(),
+            data_type: schema.column(schema.timestamp_index()).data_type,
+        };
+        let tsid_type = IndexedType {
+            idx: tsid_idx,
+            data_type: schema.column(tsid_idx).data_type,
+        };
+
+        let mut key_types = Vec::new();
+        let mut non_key_types = Vec::new();
+        for (idx, col) in schema.columns().iter().enumerate() {
+            if schema.non_key_column(idx) {
+                let _ = col
+                    .data_type
+                    .size()
+                    .context(VariableLengthType {
+                        type_str: col.data_type.to_string(),
+                    })
+                    .map_err(|e| Box::new(e) as _)
+                    .context(EncodeRecordBatch)?;
+
+                non_key_types.push(IndexedType {
+                    idx,
+                    data_type: schema.column(idx).data_type,
+                });
+            } else if idx != timestamp_type.idx && idx != tsid_idx {
+                if !matches!(col.data_type, DatumKind::String) {
+                    return StringKeyColumnRequired {
+                        type_str: col.data_type.to_string(),
+                    }
+                    .fail();
+                }
+                key_types.push(IndexedType {
+                    idx,
+                    data_type: col.data_type,
+                });
+            }
```

Comment:
I guess it will be better to organize this code block like this:
```rust
if idx == timetame_type.dix || idx == tsid_idx {
    continue
}
if schema.is_key_column(idx) {
...
} else {
...
}
```